src/dex/nerve/nerve.ts

```diff
@@ -0,0 +1,405 @@
+import { SwapSide, Network } from '../../constants';
+import {
+  wrapETH,
+  getDexKeysWithNetwork,
+  interpolate,
+  getBigIntPow,
+} from '../../utils';
+import { IDex } from '../../dex/idex';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import {
+  NerveData,
+  PoolState,
+  DexParams,
+  EventPoolMappings,
+  OptimizedNerveData,
+  NervePoolFunctions,
+  EventPoolOrMetapool,
+  NervePoolSwapParams,
+} from './types';
+import { SimpleExchange } from '../simple-exchange';
+import { NerveConfig, Adapters, NERVE_GAS_COST, NERVE_CHUNKS } from './config';
+import { NerveEventPool } from './nerve-pool';
+import _ from 'lodash';
+import { bigIntify } from './utils';
+
+export class Nerve
+  extends SimpleExchange
+  implements IDex<NerveData, DexParams, OptimizedNerveData>
+{
+  protected eventPools: EventPoolMappings = {};
+
+  readonly hasConstantPriceLargeAmounts = false;
+
+  readonly minConversionRate = '1';
+
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(NerveConfig);
+
+  logger: Logger;
+
+  static getIdentifier(dexKey: string, poolAddress: string) {
+    return `${dexKey.toLowerCase()}_${poolAddress.toLowerCase()}`;
+  }
+
+  constructor(
+    protected network: Network,
+    protected dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network] || {},
+    protected poolConfigs = NerveConfig[dexKey][network].poolConfigs,
+    protected nervePoolIface = new Interface(nervePoolABIDefault),
+  ) {
+    super(dexHelper.augustusAddress, dexHelper.provider);
+    this.logger = dexHelper.getLogger(dexKey);
+  }
+
+  get allPools() {
+    return Object.values(this.eventPools);
+  }
+
+  async initializePricing(blockNumber: number) {
+    await Promise.all(
+      Object.values(this.poolConfigs).map(poolConfig => {
+        const poolIdentifier = Nerve.getIdentifier(
+          this.dexKey,
+          poolConfig.address,
+        );
+
+        // We don't support Metapool yet
+        if (!poolConfig.isMetapool) {
+          this.eventPools[poolIdentifier] = new NerveEventPool(
+            this.dexKey,
+            this.network,
+            this.dexHelper,
+            this.logger,
+            poolConfig.name,
+          );
+          // Generate first state for the blockNumber
+          return this.eventPools[poolIdentifier].setup(blockNumber);
+        }
+      }),
+    );
+  }
+
+  getAdapters(side: SwapSide): { name: string; index: number }[] | null {
+    return this.adapters[side] || null;
+  }
+
+  async getStates(
+    pools?: EventPoolOrMetapool[],
+    blockNumber?: number,
+  ): Promise<DeepReadonly<{ state: PoolState; pool: EventPoolOrMetapool }[]>> {
+    const _pools = pools === undefined ? this.allPools : pools;
+
+    const _blockNumber =
+      blockNumber === undefined
+        ? await this.dexHelper.provider.getBlockNumber()
+        : blockNumber;
```

Comment:
if the blocknumber is not specified it is assumed it should be latest. In that case you don't need to an extra rpc call here to resolve the blocknumber. generate state already accepts `latest` as blocknumber

Reply:
A block number is needed for getState and setState...

Reply:
True